types/react/index.d.ts

```diff
@@ -914,21 +915,20 @@ declare namespace React {
     // TODO (TypeScript 3.0): <T extends unknown>
     function useRef<T>(initialValue: T): MutableRefObject<T>;
     // convenience overload for refs given as a ref prop as they typically start with a null value
+    // Note that another very common use case is for non-element refs starting with null. Thus, we 
+    // cannot enforce immutability
     /**
      * `useRef` returns a mutable ref object whose `.current` property is initialized to the passed argument
      * (`initialValue`). The returned object will persist for the full lifetime of the component.
      *
      * Note that `useRef()` is useful for more than the `ref` attribute. It’s handy for keeping any mutable
      * value around similar to how you’d use instance fields in classes.
      *
-     * Usage note: if you need the result of useRef to be directly mutable, include `| null` in the type
-     * of the generic argument.
-     *
      * @version 16.8.0
      * @see https://reactjs.org/docs/hooks-reference.html#useref
      */
     // TODO (TypeScript 3.0): <T extends unknown>
-    function useRef<T>(initialValue: T|null): RefObject<T>;
+    function useRef<T = any>(initialValue: null): MutableRefObject<T|null>;
```

Comment:
Nope, this makes the type safety _worse_ for newer typescript versions.

Reply:
See discussion below around the examples. Mutability is something we can't say anything about given modern usages